iis-workflows/iis-workflows-citationmatching-direct/src/main/java/eu/dnetlib/iis/workflows/citationmatching/direct/CitationMatchingDirectJob.java

```diff
@@ -0,0 +1,133 @@
+        conf.set("spark.kryo.registrator", "eu.dnetlib.iis.core.spark.AvroCompatibleKryoRegistrator");
+        
+        
+        
+        try (JavaSparkContext sc = new JavaSparkContext(conf)) {
+            
+            JavaRDD<ExtractedDocumentMetadataMergedWithOriginal> documentsMetadata = SparkAvroLoader.loadJavaRDD(sc, params.inputAvroPath, ExtractedDocumentMetadataMergedWithOriginal.class);
+            
+            
+            DocumentToDirectCitationMetadataConverter documentMetadataConverter = new DocumentToDirectCitationMetadataConverter();
+            JavaRDD<DocumentMetadata> simplifiedDocumentsMetadata = documentsMetadata.map(metadata -> documentMetadataConverter.convert(metadata));
+//            simplifiedDocumentsMetadata = simplifiedDocumentsMetadata.cache(); // FIXME: https://github.com/openaire/iis/issues/128
+            
+            JavaRDD<Citation> directDoiCitations = matchDoiCitations(simplifiedDocumentsMetadata);
+            
+            JavaRDD<Citation> directPmidCitations = matchPmidCitations(simplifiedDocumentsMetadata);
+            
+            JavaRDD<Citation> directCitations = mergeCitations(directDoiCitations, directPmidCitations);
+            
+            
+            
+            DirectCitationToCitationConverter directCitationToCitationConverter = new DirectCitationToCitationConverter();
+            JavaRDD<eu.dnetlib.iis.common.citations.schemas.Citation> citations = 
+                    directCitations.map(directCitation -> directCitationToCitationConverter.convert(directCitation));
+            
+            
+            SparkAvroSaver.saveJavaRDD(citations, eu.dnetlib.iis.common.citations.schemas.Citation.SCHEMA$, params.outputAvroPath);
+        }
+        
+    }
+    
+    
+    //------------------------ PRIVATE --------------------------
+    
+    
+    private static JavaRDD<Citation> matchDoiCitations(JavaRDD<DocumentMetadata> directCitationMetadata) {
+        ExternalIdCitationMatcher externalIdCitationMatcher = new ExternalIdCitationMatcher();
+        
+        return externalIdCitationMatcher.matchCitations(directCitationMetadata, "doi", x -> x.iterator().next());
+    }
+    
+    
+    private static JavaRDD<Citation> matchPmidCitations(JavaRDD<DocumentMetadata> directCitationMetadata) {
+        ExternalIdCitationMatcher externalIdCitationMatcher = new ExternalIdCitationMatcher();
+        
+        return externalIdCitationMatcher.matchCitations(directCitationMetadata, "pmid", x -> {
+            Iterator<DocumentMetadata> it = x.iterator();
+            
+            DocumentMetadata current = null;
+            while(it.hasNext()) {
+                DocumentMetadata docMeta = it.next();
+                if (StringUtils.equals(docMeta.getPublicationTypeName(), "research-article")) {
+                    return docMeta;
+                }
+                
+                current = docMeta;
+            }
+            return current;
+        });
```

Comment:
I suggest the pickSingle functions to be well defined and described external function classes.